smalld/smalld.py

```diff
@@ -108,25 +118,55 @@ def run(self):
 
             self.gateway = Gateway(gateway_url)
 
-            for data in self.gateway:
-                logger.debug("gateway payload received: %s", data)
-                for listener in self.listeners:
-                    listener(AttrDict(json.loads(data)))
+            try:
+                for data in self.gateway:
+                    logger.debug("gateway payload received: %s", data)
+                    for listener in self.listeners:
+                        listener(data)
+            except GatewayClosedException as e:
+                if e.code in recoverable_error_codes:
+                    logger.error(f"gateway closed: {e}")
+                    self.reconnect()
```

Comment:
This feels weird.

It feels like `reconnect()` closes the websocket so that we exit this loop and reconnect. So it's kind of like we're calling a method from within this loop, to exit this loop.

If we're calling this to make sure the websocket is closed, I feel like that should be done within the Gateway class

Reply:
No I think the websocket is actually closed at this point. My thinking was that this is supposed to be called to perform any tasks necessary before trying to connect again. Otherwise, its not really needed.

Reply:
yeah, I think it's not needed